add a subcommand to scrape json results from performance jobs

alaypatel07 opened the pull request:
This PR adds a command to scrape results from performance jobs. This tool works in three steps:

1. get all the data from performance job results:
```
Usage of results:
  -credentials-file string
        the credentials json file for GCS storage client
  -output-dir string
        the output directory were json data will be written (default "output/results")
  -performance-job-name string
        usuage, name of the performance job for which data is collected (default "periodic-kubevirt-e2e-k8s-1.25-sig-performance")
  -since duration
```
This will write the data out in the following format. `output-dir/performance-job-name/job-id/results.json`

2. Summarize the data into weekly runs:
```
$ perf-report-creator weekly-report --results-dir=./output/periodic-kubevirt-e2e-k8s-1.25-sig-performance --output-dir=output/weekly --vm-metrics-list=CREATE-pods-count,GET-nodes-count --vm-metrics-list=CREATE-pods-count,GET-nodes-count
```
```
Usage of weekly-report:
  -output-dir string
        the output directory were json data will be written (default "output/weekly")
  -results-dir string
        usuage, name of the performance job for which data is collected (default "output/results/periodic-kubevirt-e2e-k8s-1.25-sig-performance")
  -since duration
        Filter the periodic job in the time window (default 168h0m0s)
  -vm-metrics-list string
        comma separated list of metrics to be extracted for vms (default "vmiCreationToRunningSecondsP95")
  -vmi-metrics-list string
        comma separated list of metrics to be extracted for vmis (default "vmiCreationToRunningSecondsP95")
```

This will create a `output-dir/weekly/vmi/<week-monday-date>/<metric>/data/results.json`

3. Plot the graph from above output.
```
./perf-report-creator weekly-graph --resource=vmi --metrics-list=CREATE-pods-count,GET-nodes-count --weekly-reports-dir=./output/weekly
```

```
Usage of weekly-graph:
  -metrics-list string
        comma separated list of metrics to be plotted (default "vmiCreationToRunningSecondsP95")
  -plotly-html
        boolean for selecting what kind of graph will be plotted (default true)
  -resource string
        resource for which the graph will be plotted (default "vmi")
  -weekly-reports-dir string
        the output directory from which weekly json data will be read (default "output/weekly")
```

alaypatel07 commented:
cc @xpivarc @rthallisey 

Can I please get a first pass of this? If/When it is ready it would be great to try this against the CI gcs bucket

rthallisey commented on robots/cmd/perf-report-creator/output/vm/CREATE-pods-count/2022-10-10/data/results.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "StartDate": "2022-10-10",
+  "NumberOfDays": 0,
```
Is this an example of real data you scraped?

alaypatel07 replied:
no @rthallisey these are made up numbers to demo the output directory structure.

rthallisey commented on robots/cmd/perf-report-creator/output/vm/CREATE-pods-count/2022-10-10/data/results.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "StartDate": "2022-10-10",
```
How can I tell this data is weekly?

rthallisey replied:
Should we consider a different path so we have room for expansion?  `../2022-10-10/data/results.json` makes it seem like we're tracking daily.

alaypatel07 replied:
@rthallisey I had a bug in there where the `NumberOfDays` was not being updated, I fixed it and pushed a commit.

But the output-dir flag can help here `-output-dir output/weekly` will give you a separate directory for weekly collection

alaypatel07 replied:
@rthallisey I have copied this from one of the perf jobs  `robots/cmd/perf-report-creator/test-build-log.txt` to make sure the unit test for regex match are finding the right values that file has the real data. Going further @dhiller mentioned that we should modify the performance job to write the output to a separate file. but unfortunately this is the only way to get historical data

alaypatel07 replied:
addressed the original question by separating this into two, one to scrap the results from the perf job and second to aggregate it into a weekly format

rthallisey commented:
cc @dhiller

dhiller commented on robots/cmd/perf-report-creator/main.go:
```diff
@@ -0,0 +1,319 @@
+		// return fmt.Errorf("Failed to create new storage client: %v.\n", err)
+	}
+
+	allJobs, err := listAllJobs(ctx, storageClient)
```
And looking further into this, it seems like you assume that the content will be traversed recursively which I think is not the case. I think you need to descend into each directory that you want to traverse.

Or my gcs lib knowledge is not up to date, which may well be :smile:

alaypatel07 replied:
yes, I need to update the url to `gs://kubevirt-prow/logs/periodic-kubevirt-e2e-k8s-1.25-sig-performance/` I'll push an update for that with a todo to improve the performance of the `since` part

Although I do make a subsequent call to download the build-log.txt at

```
gs://kubevirt-prow/logs/periodic-kubevirt-e2e-k8s-1.25-sig-performance/<job-id>/build-log.txt
```

Does the above appropriately handle what is expected from the api?
> And looking further into this, it seems like you assume that the content will be traversed recursively which I think is not the case. I think you need to descend into each directory that you want to traverse.

alaypatel07 replied:
@dhiller pushed a fix as suggested with a todo for sorting enhancement. PTAL

alaypatel07 commented:
Added a mechanism to plot the weekly graphs:
```
 ./perf-report-creator weekly-graph --resource vmi --metric CREATE-pods-count --weekly-reports-dir output/weekly --plotly-html=false
 ```
![image](https://user-images.githubusercontent.com/8192650/226339215-7f2e2ca7-4d8b-46f1-8b39-1d9419c775f2.png)

Option for more dynamic graph:
```
 ./perf-report-creator weekly-graph --resource vmi --metric CREATE-pods-count --weekly-reports-dir output/weekly --plotly-html=true
 ```
![image](https://user-images.githubusercontent.com/8192650/226339315-8e2e9f73-a372-4703-92c5-0658e625fba6.png)

alaypatel07 commented:
@rthallisey @dhiller 

guys can I please get another pass on this PR?

One feedback I received was that we want to be selective of the metric we put on disk so that the data is bound on certain limit. 

I have not hard-coded the filter, but you can imagine a job that runs phase 1, phase 2(with selected metric), phase 3(with selected metric) and upload the weekly data and graph to github repo.

@rthallisey  does the above work for limiting the number of metrics collected?

xpivarc reviewed (commented):
Please remove `test-build-log.txt`, and `output/weekly` if they are not necessary.

xpivarc commented:
@brianmcarey @dhiller 
Can we merge this as the first iteration? The next step would be to implement necessary improvements so we can run this periodically and we then create a job for this.

alaypatel07 commented on robots/cmd/perf-report-creator/main.go:
```diff
@@ -0,0 +1,319 @@
+
+func flagOptions() options {
+	o := options{}
+	flag.DurationVar(&o.since, "since", 24*7*time.Hour, "Filter the periodic job in the time window")
```
updated.

alaypatel07 commented:
@xpivarc @rthallisey 

pushed the following changes:
1. added README.md
2. updated defaults to 24 hours
3. removed tests files, to be added later

brianmcarey commented on robots/cmd/perf-report-creator/main.go:
```diff
@@ -0,0 +1,453 @@
+			errs = append(errs, err)
+		}
+
+		//d := creationTime.Format("2006-01-02T00:00:00Z00:00")
```
Is this a leftover?

brianmcarey commented on robots/cmd/perf-report-creator/main.go:
```diff
@@ -0,0 +1,453 @@
+		log.Printf("job: %s, error running readLinesAndMatchRegex. %+v\n", jobID, err)
+		return Result{}, err
+	}
+	//lines = lines[3:]
```
leftover?

brianmcarey reviewed (commented):
Looks good @alaypatel07  - just a couple of nit questions on some comments.

alaypatel07 commented:
@brianmcarey updated as requested. PTAL

brianmcarey commented:
/ok-to-test

brianmcarey reviewed (approved):
/approve

Thanks @alaypatel07 !
